zeebe/benchmarks/project/src/main/java/io/camunda/zeebe/ProcessInstanceMonitor.java

```diff
@@ -0,0 +1,300 @@
+            .log("Failed to observe process instance after {} seconds");
+        latency.record(MONITOR_TIMEOUT);
+        return;
+      }
+
+      final var requestTime = Instant.now();
+      client
+          .getProcessInstance(processInstanceKey)
+          .whenCompleteAsync((r, e) -> onProcessInstanceResponse(requestTime, e), executor);
+    }
+
+    private void onProcessInstanceResponse(final Instant requestTime, final Throwable error) {
+      final TaskState taskState = error != null ? handleError(error) : handleSuccess(requestTime);
+
+      if (taskState == TaskState.STOP) {
+        monitoringCount.decrementAndGet();
+        return;
+      }
+
+      idleStrategy.idle(0);
+      requestCount++;
+      withContext(LOGGER.atTrace()).log("Retrying process instance monitoring...");
+      executor.execute(this);
+    }
+
+    private TaskState handleSuccess(final Instant requestTime) {
+      // subtract at least one backoff delay to account for a possible error margin of having waited
+      // up to BACKOFF_DELAY too long in the last cycle
+      final var piLatency = Duration.between(startTime, requestTime).minus(BACKOFF_DELAY);
+      latency.record(piLatency);
+      withContext(LOGGER.atDebug())
+          .addArgument(piLatency)
+          .log("Process instance was visible after {}");
+
+      return TaskState.STOP;
+    }
+
+    private TaskState handleError(final Throwable error) {
+      withContext(LOGGER.atTrace())
+          .setCause(error)
+          .log("Failed to search process instances to measure visibility latency");
+
+      // we expect 404 (not found) until it is; however, for other errors, we will likely have
+      // skewed measurements, so we should stop monitoring that PI
+      if (error instanceof HttpResponseException problem && problem.code() == 404) {
+        return TaskState.RETRY;
+      }
+
+      withContext(LOGGER.atDebug())
+          .setCause(error)
+          .log("Unexpected error while monitoring PI, will ignore to avoid skewed measurements");
+      errors.increment();
+      return TaskState.STOP;
+    }
+
+    private LoggingEventBuilder withContext(final LoggingEventBuilder builder) {
+      return builder
+          .addKeyValue(PROCESS_INSTANCE_KEY.key(), processInstanceKey)
+          .addKeyValue(START_TIME.key(), startTime)
+          .addKeyValue(REQUEST_COUNT.key(), requestCount)
+          .addKeyValue(MONITORING_COUNT.key(), monitoringCount.get());
+    }
+
+    private enum TaskState {
+      STOP,
+      RETRY;
+    }
+  }
+
+  enum LoggingKeys {
+    REQUEST_COUNT("requestCount"),
+    PROCESS_INSTANCE_KEY("processInstanceKey"),
+    START_TIME("startTime"),
+    MONITORING_COUNT("monitoringCount");
+
+    private final String key;
+
+    LoggingKeys(final String key) {
+      this.key = key.intern();
+    }
+
+    String key() {
+      return key;
+    }
+  }
+
+  @SuppressWarnings("NullableProblems")
+  public enum MetricsDoc implements ExtendedMeterDocumentation {
```

Comment:
🔧 In general, I understand the reasoning behind it (and also sometimes like it that we have the necessary coupled data together), but I feel it is often a bit overengineered, especially for nonprod code. But it might just be me.

Reply:
I think if we rely on it, it makes sense. I know it seems like non-prod because it's not meant for users, but I see it as becoming an important way to measure performance of the data pipeline. I know we could say, let's start small, I guess I don't trust that it will be changed later due to inertia :shrug: 

I can extract it top level though so it's ready to accept more client side metrics in the future, and not be super specific to just API monitoring.